### PR TITLE
Fix parsing of final states

### PR DESCRIPTION
**Pull request by jurajsic**
This PR fixes parsing into NFA and AFA when final states are given as a conjuction of negated non-final states. There was a problem if there was only one such negated state for NFA, and for AFA it was completely ignored. Also it fixes parsing into NFA when final formula is given as true/false (then final states are either empty set or all states).

### CODE REVIEW

File: src/inter-aut.cc [hunk 1]
```diff
@@ -601,7 +599,11 @@ bool Mata::IntermediateAut::is_graph_conjunction_of_negations(const Mata::Formul
             return false;
     }
 
-    return true;
+    if (act_graph->node.is_operator() && act_graph->node.is_neg()) {
```
Comment (tfiedor) [inline src/inter-aut.cc]: I Believe this could be shortened to `return act_graph->node.is_operator() && act_graph->node.is_neg()`

**Review by tfiedor — APPROVED** [review]
Seems ok

File: src/inter-aut.cc (hunk 1 above)
Comment (jurajsic) [reply src/inter-aut.cc]: Yeah that's true, fixed

File: src/nfa/nfa.cc [hunk 2]
```diff
@@ -1419,6 +1411,26 @@ Nfa Mata::Nfa::construct(
         aut.delta.add(src_state, symbol, tgt_state);
     }
 
+    std::unordered_set<std::string> final_formula_nodes;
+    if (!(inter_aut.final_formula.node.name == "true" || inter_aut.final_formula.node.name == "false")) { 
+        final_formula_nodes = inter_aut.final_formula.collect_node_names();
+    }
+    bool final_nodes_are_negated = (inter_aut.final_formula.node.name == "true" || inter_aut.are_final_states_conjunction_of_negation());
```
Comment (martinhruska) [inline src/nfa/nfa.cc]: We should discuss whether this should not be a function in inter_automaton, something like `inter_aut.final_nodes_are_negated()`. The same for condition above.

File: src/afa/afa.cc [hunk 3]
```diff
@@ -1121,6 +1115,24 @@ Afa Mata::Afa::construct(
                       create_node(curr_graph->collect_node_names()));
     }
 
+	// TODO final states can be also given as true/false
+	if (inter_aut.are_final_states_conjunction_of_negation()) {
+		// final states are given as a conjunction of non-final states
```
Comment (martinhruska) [inline src/afa/afa.cc:1114]: Maybe you can encapsulate this to intermediate automaton class to make code here cleaner.

**Review by martinhruska — CHANGES_REQUESTED** [review]

**Comment by martinhruska** [comment]
Please check my comment, if they don't make sense to you, I'm gonna approve the PR to not delay merge.

**Review by Adda0 — APPROVED** [review]
I support Martin's suggestions about extracting some operations into separate methods.

Otherwise, the changes seem OK to me.

File: src/nfa/nfa.cc (hunk 2 above)
Comment (jurajsic) [reply src/nfa/nfa.cc]: I don't think this should be a function of inter_automaton, I am asking if I have conjunction of negations or the constant true, function call would just make it less obvious.
The condition `inter_aut.final_formula.node.name == "true" || inter_aut.final_formula.node.name == "false"` should probably be a function of inter_automaton, or better, function of FormulaGraph, something like `is_constant`, I'll add it. A cleaner solution would maybe be adding constants to FormulaGraph.

File: src/afa/afa.cc (hunk 3 above)
Comment (jurajsic) [reply src/afa/afa.cc:1114]: This was just a quick fix so it parses final states correctly (and there is still the need to fix true/false). To make it a method of intermediate automaton, I would need to get states from the transition relation and I am not sure how exactly it looks for AFA.

**Review by martinhruska — APPROVED** [review]

File: src/nfa/nfa.cc (hunk 2 above)
Comment (martinhruska) [reply src/nfa/nfa.cc]: Yes, it would be better to modify Node and add an option to enum that the node is a constant operand true/false.

**Comment by Adda0** [comment]
@jurajsic Are you done with this PR? Can it be merged?

File: src/nfa/nfa.cc (hunk 2 above)
Comment (jurajsic) [reply src/nfa/nfa.cc]: I added an issue #174 for this

**Comment by jurajsic** [comment]
I tried to solve issue #175 in the reverted commit, but it resulted in failing tests, so I just left some stuff from the commit which worked correctly. If nobody complains, I will merge it tomorrow morning.

**Comment by Adda0** [comment]
Unless you do not need these changes immediately, give us at least tomorrow to review the new changes. If you do need them, tell us and we will quicken our reviews, but give us some time either way.

**Comment by jurajsic** [comment]
Ok, I don't need them immediately.

**Review by Adda0 — APPROVED** [review]
The changes look good to me.

File: include/mata/nfa.hh [hunk 4]
```diff
@@ -992,31 +992,15 @@ Nfa construct(
...
         StringToStateMap*                    state_map = nullptr)
 { // {{{
-    Nfa aut;
-
-    bool remove_symbol_map = false;
-    if (nullptr == symbol_map)
-    {
-        symbol_map = new StringToSymbolMap();
-        remove_symbol_map = true;
+    StringToSymbolMap tmp_symbol_map;
```
Comment (martinhruska) [inline include/mata/nfa.hh:995]: Why do you need this temp thing here?

**Review by martinhruska — APPROVED** [review]

**Comment by martinhruska** [comment]
I went through the new stuff quickly and it looks fine. The only things I don't understand are that `tmp_*` maps.

**Comment by martinhruska** [comment]
I went through the new stuff quickly and it looks fine. The only things I don't understand are that `tmp_*` maps.

File: include/mata/nfa.hh (hunk 4 above)
Comment (jurajsic) [reply include/mata/nfa.hh:995]: It's not needed, but it's safer than using new/delete.

**Pull request merged**